src/main/java/org/apache/sling/scripting/core/impl/jsr223/SlingScriptEngineManager.java

```diff
@@ -0,0 +1,320 @@
+        } finally {
+            readWriteLock.writeLock().unlock();
+        }
+    }
+
+    @Override
+    public void bundleChanged(BundleEvent event) {
+        if (event.getType() == BundleEvent.STARTED
+                && event.getBundle().getEntry(ENGINE_FACTORY_SERVICE) != null) {
+            synchronized (this.engineSpiBundles) {
+                this.engineSpiBundles.add(event.getBundle());
+            }
+            updateFactories();
+        } else if (event.getType() == BundleEvent.STOPPED) {
+            boolean refresh;
+            synchronized (this.engineSpiBundles) {
+                refresh = this.engineSpiBundles.remove(event.getBundle());
+            }
+            if (refresh) {
+                updateFactories();
+            }
+        }
+    }
+
+    public Map<String, Object> getProperties(ScriptEngineFactory factory) {
+        readWriteLock.readLock().lock();
+        try {
+            return factoriesProperties.get(factory);
+        } finally {
+            readWriteLock.readLock().unlock();
+        }
+    }
+
+    @Activate
+    private void activate(ComponentContext componentContext) {
+        this.componentContext = componentContext;
+        updateFactories();
+    }
+
+    private void bindScriptEngineFactory(final ServiceReference<ScriptEngineFactory> serviceReference,
+                                           final ScriptEngineFactory factory) {
+        synchronized (this.serviceReferences) {
+            serviceReferences.add(serviceReference);
+        }
+        updateFactories();
+        postEvent(SlingScriptConstants.TOPIC_SCRIPT_ENGINE_FACTORY_ADDED, factory);
+    }
+
+    private void unbindScriptEngineFactory(final ServiceReference<ScriptEngineFactory> serviceReference, final ScriptEngineFactory
+            factory) {
+        synchronized (this.serviceReferences) {
+            serviceReferences.remove(serviceReference);
+            if (componentContext != null) {
+                componentContext.getBundleContext().ungetService(serviceReference);
+            }
+        }
+        updateFactories();
+        postEvent(SlingScriptConstants.TOPIC_SCRIPT_ENGINE_FACTORY_REMOVED, factory);
+    }
+
+    private void updateFactories() {
+        readWriteLock.writeLock().lock();
+        try {
+            factories = new TreeSet<>();
+            internalManager = new ScriptEngineManager(SlingScriptEngineManager.class.getClassLoader());
+            long fakeBundleIdCounter = Long.MIN_VALUE;
+            // first add the platform factories
+            for (ScriptEngineFactory factory : internalManager.getEngineFactories()) {
+                SortableScriptEngineFactory ssef = new SortableScriptEngineFactory(factory, fakeBundleIdCounter++, 0);
+                factories.add(ssef);
+            }
+            // then factories from SPI Bundles
+            for (Bundle bundle : engineSpiBundles) {
+                URL url = bundle.getEntry(ENGINE_FACTORY_SERVICE);
```

Comment:
Can't you rely on https://docs.oracle.com/javase/9/docs/api/java/util/ServiceLoader.html?

Reply:
We probably could - feel free to push a commit changing the code. :)

I didn't want to change more code than needed - this method is copied from the old `ScriptEngineManagerFactory` (https://github.com/apache/sling-org-apache-sling-scripting-core/pull/1/files#diff-592cfbf344b2569e240757d7fb09e125L119).

Reply:
That would be problematic as the serviceloader would see too much and too little at the same time - i think. At this point, we really only want the engines in that bundle.

Reply:
Well, I guess it will work llike this:

```
-            // then factories from SPI Bundles
-            for (Bundle bundle : engineSpiBundles) {
-                URL url = bundle.getEntry(ENGINE_FACTORY_SERVICE);
-                InputStream ins = null;
-                try {
-                    ins = url.openStream();
-                    BufferedReader reader = new BufferedReader(new InputStreamReader(ins, StandardCharsets.UTF_8));
-                    String line;
-                    while ((line = reader.readLine()) != null) {
-                        if (!line.startsWith("#") && line.trim().length() > 0) {
-                            try {
-                                Class<ScriptEngineFactory> clazz = (Class<ScriptEngineFactory>) bundle.loadClass(line);
-                                SortableScriptEngineFactory ssef = new SortableScriptEngineFactory(clazz.getDeclaredConstructor()
-                                        .newInstance(), fakeBundleIdCounter++, 0);
-                                factories.add(ssef);
-                            } catch (Throwable t) {
-                                LOG.error("Cannot register ScriptEngineFactory " + line, t);
-                            }
-                        }
-                    }
-                } catch (IOException ioe) {
-                    LOG.error("Unable to process bundle " + bundle.getSymbolicName(), ioe);
-                } finally {
-                    if (ins != null) {
-                        try {
-                            ins.close();
-                        } catch (IOException ioe) {
-                            LOG.error("Unable to release stream resource.", ioe);
+            ClassLoader loader = Thread.currentThread().getContextClassLoader();
+            try {
+                Thread.currentThread().setContextClassLoader(null);
+                // then factories from SPI Bundles
+                for (Bundle bundle : engineSpiBundles) {
+                    try {
+                        ScriptEngineManager manager = new ScriptEngineManager(bundle.adapt(BundleWiring.class).getClassLoader());
+                        for (ScriptEngineFactory factory : manager.getEngineFactories()) {
+                            factories.add(new SortableScriptEngineFactory(factory, fakeBundleIdCounter++, 0));
                         }
+                    } catch (Exception ex) {
+                        LOG.error("Unable to process bundle " + bundle.getSymbolicName(), ex);
                     }
                 }
             }
+            finally {
+                Thread.currentThread().setContextClassLoader(loader);
+            }
```